packages/nuxt/src/app/nuxt.ts

```diff
@@ -187,6 +187,17 @@ export function createNuxtApp (options: CreateOptions) {
 
   nuxtApp.hooks = createHooks<RuntimeNuxtHooks>()
   nuxtApp.hook = nuxtApp.hooks.hook
+
+  if (process.server) {
+    async function contextCaller (hooks: HookCallback[], args: any[]) {
+      for (const hook of hooks) {
+        await nuxtAppCtx.call(nuxtApp, () => hook(...args))
+      }
+    }
+    // Patch callHook to preserve NuxtApp context on server
+    nuxtApp.hooks.callHook = (name: any, ...args: any[]) => nuxtApp.hooks.callHookWith(contextCaller, name, ...args)
```

Comment:
For followup, we could support `setCaller` for hookable library https://github.com/unjs/hookable/issues/74

Reply:
```suggestion
    // TODO: Refactor after https://github.com/unjs/hookable/issues/74
    nuxtApp.hooks.callHook = (name: any, ...args: any[]) => nuxtApp.hooks.callHookWith(contextCaller, name, ...args)
```